kafka-persistence/src/main/scala/com/evolutiongaming/kafka/flow/kafkapersistence/KafkaPersistence.scala

```diff
@@ -0,0 +1,167 @@
+import com.evolutiongaming.kafka.journal.util.SkafkaHelper._
+import com.evolutiongaming.skafka.consumer.{ConsumerConfig, ConsumerOf, ConsumerRecord, WithSize}
+import com.evolutiongaming.skafka.producer.{Producer, ProducerRecord}
+import com.evolutiongaming.skafka.{FromBytes, Offset, Partition, ToBytes, Topic, TopicPartition}
+import com.evolutiongaming.sstream.Stream
+import com.olegpy.meow.effects._
+import monocle.macros.GenLens
+import scodec.bits.ByteVector
+
+import scala.concurrent.duration._
+import scala.util.control.NoStackTrace
+
+final case class KafkaPersistence[F[_], K, S, A](
+                                                  keysOf: KeysOf[F, K],
+                                                  snapshots: SnapshotPersistenceOf[F, K, S, A],
+                                                  onRecoveryFinished: F[Unit]
+                                                )
+
+object KafkaPersistence {
+
+  private type Record = ConsumerRecord[String, ByteVector]
+  private type BytesByKey = Map[String, ByteVector]
+
+  private object BytesByKey {
+    val zero: BytesByKey = Map.empty
+  }
+
+  private case class MissingOffsetError(topicPartition: TopicPartition) extends NoStackTrace
+
+  final case class Of[F[_], K, S, A](create: Partition => F[KafkaPersistence[F, K, S, A]])
+
+  object Of {
+    def apply[F[_] : BracketThrowable : Producer : ToBytes[*[_], S] : FromBytes[*[_], S] : FromTry : Log : Sync, S, A](
+                                                                                                                        consumerOf: ConsumerOf[F],
+                                                                                                                        consumerConfig: ConsumerConfig,
+                                                                                                                        stateTopic: Topic
+                                                                                                                      ): Of[F, KafkaKey, S, A] =
+      this { partition =>
+        for {
+          stateData <- readState(
+            consumerOf = consumerOf,
+            consumerConfig = consumerConfig,
+            stateTopic = stateTopic,
+            partition = partition
+          )
+          stateRef <- Ref.of(stateData)
+        } yield stateRef.runState { implicit state =>
+          KafkaPersistence[F, S, A](stateTopic)
+        }
+      }
+  }
+
+  def apply[F[_] : Producer : Monad : FromTry : Sync : Log : MonadState[*[_], BytesByKey] : FromBytes[*[_], S] : ToBytes[
+    *[_],
+    S
+  ], S, A](stateTopic: Topic): KafkaPersistence[F, KafkaKey, S, A] = {
+    val snapshotsOf = new SnapshotsOf[F, KafkaKey, S] {
+      override def apply(key: KafkaKey): F[Snapshots[F, S]] =
+        Snapshots.of(
+          key,
+          new SnapshotDatabase[F, KafkaKey, S] {
+            override def get(key: KafkaKey) =
+              for {
+                state <- MonadState.get
```

Comment:
Why use `MonadState` if you have `Sync` anyway? `Ref` could be easier and more performant option here.

Reply:
Actually, when I look at the code, there is a point of `MonadState`, because there is a plenty of unit-testable logic here. But then why `Sync`?

Reply:
I need Sync here because I directly call `Snapshots.of` which asks `Sync`, and I probably shouldn't be doing it. Let me see if I can get rid of it.

Reply:
Dropped `Sync`,  used externally provided factory for state containers.